awx/main/tasks.py

```diff
@@ -602,26 +602,25 @@ def update_inventory_computed_fields(inventory_id, should_update_hosts=True):
 
 
 def update_smart_memberships_for_inventory(smart_inventory):
-    with advisory_lock('update_smart_memberships_for_inventory-{}'.format(smart_inventory.id)):
-        current = set(SmartInventoryMembership.objects.filter(inventory=smart_inventory).values_list('host_id', flat=True))
-        new = set(smart_inventory.hosts.values_list('id', flat=True))
-        additions = new - current
-        removals = current - new
-        if additions or removals:
-            with transaction.atomic():
-                if removals:
-                    SmartInventoryMembership.objects.filter(inventory=smart_inventory, host_id__in=removals).delete()
-                if additions:
-                    add_for_inventory = [
-                        SmartInventoryMembership(inventory_id=smart_inventory.id, host_id=host_id)
-                        for host_id in additions
-                    ]
-                    SmartInventoryMembership.objects.bulk_create(add_for_inventory)
-            logger.debug('Smart host membership cached for {}, {} additions, {} removals, {} total count.'.format(
-                smart_inventory.pk, len(additions), len(removals), len(new)
-            ))
-            return True  # changed
-        return False
+    current = set(SmartInventoryMembership.objects.filter(inventory=smart_inventory).values_list('host_id', flat=True))
+    new = set(smart_inventory.hosts.values_list('id', flat=True))
+    additions = new - current
+    removals = current - new
+    if additions or removals:
+        with transaction.atomic():
+            if removals:
+                SmartInventoryMembership.objects.filter(inventory=smart_inventory, host_id__in=removals).delete()
+            if additions:
+                add_for_inventory = [
+                    SmartInventoryMembership(inventory_id=smart_inventory.id, host_id=host_id)
+                    for host_id in additions
+                ]
+                SmartInventoryMembership.objects.bulk_create(add_for_inventory, ignore_conflicts=True)
```

Comment:
We added this lock in the first place because processes that modify the same inventory/host simultaneously would perform this same calculation and potentially generate `IntegrityError`s if these bulk creates run at the same time.

```python
Traceback (most recent call last): File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute return self.cursor.execute(sql, params) psycopg2.IntegrityError: duplicate key value violates unique constraint "main_smartinventorymembe_host_id_inventory_id_58137be6_uniq" DETAIL: Key (host_id, inventory_id)=(15410, 17) already exists. The above exception was the direct cause of the following exception:

Traceback (most recent call last): File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1152, in run self.pre_run_hook(self.instance)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1647, in pre_run_hook update_smart_memberships_for_inventory(job.inventory)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 626, in update_smart_memberships_for_inventory SmartInventoryMembership.objects.bulk_create(add_for_inventory)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/models/query.py", line 443, in bulk_create ids = self._batched_insert(objs_without_pk, fields, batch_size)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/models/query.py", line 1096, in _batched_insert inserted_id = self._insert(item, fields=fields, using=self.db, return_id=True)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/models/query.py", line 1079, in _insert return query.get_compiler(using=using).execute_sql(return_id)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/models/sql/compiler.py", line 1112, in execute_sql cursor.execute(sql, params)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute return self.cursor.execute(sql, params)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/utils.py", line 94, in __exit__ six.reraise(dj_exc_type, dj_exc_value, traceback)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/utils/six.py", line 685, in reraise raise value.with_traceback(tb)
File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute return self.cursor.execute(sql, params) django.db.utils.IntegrityError: duplicate key value violates unique constraint "main_smartinventorymembe_host_id_inventory_id_58137be6_uniq" DETAIL: Key (host_id, inventory_id)=(15410, 17) already exists.
```